backend/server.js

```diff
@@ -213,6 +216,111 @@ sbp('sbp/selectors/register', {
+    } catch (e) {
+      console.warn(e, `Error parsing manifest for ${cid}. It's probably not a file manifest.`)
+      throw new BackendErrorNotFound()
+    }
+    // The keys to be deleted are not read from or updated, so they can be deleted
+    // without using a queue
+    const resourcesKey = `_private_resources_${owner}`
+    await removeFromIndexFactory(resourcesKey)(cid)
+
+    await sbp('chelonia.db/delete', `_private_owner_${cid}`)
+    await sbp('chelonia.db/delete', `_private_size_${cid}`)
+    await sbp('chelonia.db/delete', `_private_deletionToken_${cid}`)
+
+    await sbp('chelonia.db/set', cid, '')
+  },
+  // eslint-disable-next-line require-await
+  async 'backend/deleteContract' (cid: string): Promise<void> {
+    let contractsPendingDeletion = sbp('okTurtles.data/get', 'contractsPendingDeletion')
+    if (!contractsPendingDeletion) {
+      contractsPendingDeletion = new Set()
+      sbp('okTurtles.data/set', 'contractsPendingDeletion', contractsPendingDeletion)
+    }
+    // Avoid deadlocks due to loops
+    if (contractsPendingDeletion.has(cid)) {
+      return
+    }
+    contractsPendingDeletion.add(cid)
+
+    return sbp('chelonia/queueInvocation', cid, async () => {
+      const owner = await sbp('chelonia.db/get', `_private_owner_${cid}`)
+      const rawManifest = await sbp('chelonia.db/get', cid)
+      if (rawManifest === '') throw new BackendErrorGone()
+      if (!rawManifest) throw new BackendErrorNotFound()
+
+      const resourcesKey = `_private_resources_${cid}`
+      const resources = await sbp('chelonia.db/get', resourcesKey)
+      if (resources) {
+        await Promise.allSettled(resources.split('\x00').map((resourceCid) => {
+          const parsed = parseCID(resourceCid)
+
+          if (parsed.code === multicodes.SHELTER_CONTRACT_DATA) {
+            return sbp('chelonia.persistentActions/enqueue', ['backend/deleteContract', resourceCid])
+          } else if (parsed.code === multicodes.SHELTER_FILE_MANIFEST) {
+            return sbp('chelonia.persistentActions/enqueue', ['backend/deleteFile', resourceCid])
+          }
+
+          return undefined
+        }))
+      }
+      await sbp('chelonia.db/delete', resourcesKey)
+
+      const latestHEADinfo = await sbp('chelonia/db/latestHEADinfo', cid)
+      if (latestHEADinfo) {
+        for (let i = latestHEADinfo.height; i > 0; i--) {
+          const eventKey = `_private_hidx=${cid}#${i}`
+          const event = await sbp('chelonia.db/get', eventKey)
+          if (event) {
+            await sbp('chelonia.db/delete', event)
+            await sbp('chelonia.db/delete', eventKey)
+          }
+        }
+        await sbp('chelonia/db/deleteLatestHEADinfo', cid)
+      }
+
+      const kvIndexKey = `_private_kvIdx_${cid}`
+      const kvKeys = await sbp('chelonia.db/get', kvIndexKey)
+      if (kvKeys) {
+        await kvKeys.split('\x00').map((key) => {
+          return sbp('chelonia.db/delete', `_private_kv_${cid}_${key}`)
+        })
```

Comment:
Think this is missing a surrounding `Promise.all` or `Promise.allSettled`?

Reply:
Indeed.